### PR TITLE
Fixes #259 - Implements  `is_first_version` and `is_last_version` attributes

### LINKED ISSUE #259: Scripts API enhancements
Please expose the following in the scripts API:

- The #show endpoint
- The `executable_settings` field and (if #258 is a go) the `executable_settings_filename` field
- and _discuss_: should `/filter` return all versions of a script? It currently does.
  - personal preference is to not allow this... but the corollary is how do we allow seeing older versions?

### PR DESCRIPTION
**Pull request by atruskie**
Adds virtual fields `is_first_version` and `is_last_version` to the
Scripts API. Does NOT implement the `/groups` API endpoint - unnecessary
at this time.

Note one unit test currently fails because #286 is not implemented.

### CODE REVIEW

File: app/models/script.rb [hunk 1]
```diff
@@ -27,51 +27,103 @@ def latest_version
...
+    joins(query).select('"scripts".*, "s2"."last_version", "s2"."first_version"')
+  end
+
+  # a fake reference to the aliased table defined in the default scope
+  @script_group = Arel::Table.new('s2', @arel_engine)
+
+  # HACK: reload item - this is a poor version of Persistence::reload
+  # https://github.com/rails/rails/blob/1f98eb60e59f4f70ef66ac2454ad029f46e3b27c/activerecord/lib/active_record/persistence.rb#L432
+  # The only difference is the `unscoped` method is removed.
+  def reload(options = nil)
```
Comment (cofiem) [inline app/models/script.rb]: why is this needed?

Comment (atruskie) [inline app/models/script.rb]: See:
- https://github.com/QutBioacoustics/baw-server/pull/287/files/4db980b4531b65d47e730247b0e0c6b7067ec239#diff-9cacd63540753d6929377d6cb58210cbR157
- https://github.com/QutBioacoustics/baw-server/pull/287/files/4db980b4531b65d47e730247b0e0c6b7067ec239#diff-42b71321a35f19e8409b3b39966e6f24R62
- https://github.com/QutBioacoustics/baw-server/pull/287/files/4db980b4531b65d47e730247b0e0c6b7067ec239#diff-42b71321a35f19e8409b3b39966e6f24R69

File: spec/acceptance/scripts_spec.rb [hunk 2]
```diff
@@ -119,11 +210,38 @@ def scripts_id_param
...
         ],
-        data_item_count: 1,
+        data_item_count: 4,
         response_body_content: [
             '"analysis_identifier":{"contains":" identifier "}',
             '"executable_settings":"']
     })
   end
 
+  post '/scripts/filter' do
```
Comment (cofiem) [inline spec/acceptance/scripts_spec.rb]: Is this the test that fails due to #286?

Comment (atruskie) [inline spec/acceptance/scripts_spec.rb]: Nope, see https://github.com/QutBioacoustics/baw-server/pull/287/files/4db980b4531b65d47e730247b0e0c6b7067ec239#diff-944a313f082f8ee3273a4aea8688a805R73

**Comment by atruskie** [comment]
So I'm reviewing the changes in 0ab8f91.

Converting the `last_version` and `first_version` properties to queries only is pretty bad for performance. A `GET /scripts` goes from 1 database query to **2×N + 1** queries - and that's ignoring the paging query! Essentially, the values in the default scope never get used.

Please consider the changes in 8bc7cb2 which use the attribute values if they are present and query for them in the rare case they are not present.

**Comment by cofiem** [comment]
That looks reasonable - have you tested your most recent change much?

**Comment by atruskie** [comment]
Unit and integration tests pass. Basic tests with http endpoint tester behave as I expect for `index`, `show`, and `filter` (with a `is_last_version` filter). `index` also works with `?filter_is_last_version=true` usecase.

**Comment by cofiem** [comment]
cool, I'll review and merge tonight
